File: common/src/client/java/juuxel/adorn/client/book/BookManager.java
package juuxel.adorn.client.book;

import juuxel.adorn.AdornCommon;
import juuxel.adorn.item.AdornBookItem;
import net.minecraft.ChatFormatting;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.FileToIdConverter;
import net.minecraft.resources.Identifier;
import net.minecraft.server.packs.resources.ResourceManager;
import net.minecraft.server.packs.resources.SimpleJsonResourceReloadListener;
import net.minecraft.util.profiling.ProfilerFiller;

import java.util.Map;
import java.util.function.Consumer;

public final class BookManager extends SimpleJsonResourceReloadListener<Book> implements AdornBookItem.BookTooltipProvider {
    public static final Identifier ID = AdornCommon.id("book_manager");
    public static final BookManager INSTANCE = new BookManager();
    public static final String DATA_TYPE = "adorn/books";

    private Map<Identifier, Book> books = Map.of();

    private BookManager() {
        super(Book.CODEC, FileToIdConverter.json(DATA_TYPE));
    }

    @Override
    protected void apply(Map<Identifier, Book> prepared, ResourceManager manager, ProfilerFiller profiler) {
        books = Map.copyOf(prepared);
    }

    public boolean contains(Identifier id) {
        return books.containsKey(id);
    }

    public Book get(Identifier id) {
        var book = books.get(id);
        if (book == null) {
            throw new IllegalArgumentException("Tried to get unknown book '%s' from BookManager".formatted(id));
        }
        return book;
    }

    @Override
    public void appendTooltip(Identifier bookId, Consumer<Component> builder) {
        if (contains(bookId)) {
           builder.accept(Component.translatable("book.byAuthor", get(bookId).author()).withStyle(ChatFormatting.GRAY));
        }
    }

    public static void setupTooltipProvider() {
        AdornBookItem.tooltipProvider = INSTANCE;
    }
}

File: fabric/src/main/kotlin/juuxel/adorn/client/resources/BookManager.kt
package juuxel.adorn.client.resources

import blue.endless.jankson.JsonObject
import juuxel.adorn.AdornCommon
import juuxel.adorn.client.book.Book
import juuxel.adorn.resources.Json5DataLoader
import net.fabricmc.api.EnvType
import net.fabricmc.api.Environment
import net.fabricmc.fabric.api.resource.IdentifiableResourceReloadListener
import net.minecraft.resource.ResourceManager
import net.minecraft.util.Identifier
import net.minecraft.util.profiler.Profiler

@Environment(EnvType.CLIENT)
object BookManager : Json5DataLoader(AdornCommon.id("book_manager"), "adorn/books", Book.jankson().build()), IdentifiableResourceReloadListener {
    private const val PREFIX_LENGTH = "adorn_books".length
    private const val SUFFIX_LENGTH = ".json5".length
    private var books: Map<Identifier, Book> = emptyMap()

    override fun apply(jsons: Map<Identifier, JsonObject>, manager: ResourceManager, profiler: Profiler) {
        books = jsons.asSequence()
            .map { (key, value) ->
                Identifier(key.namespace, key.path.substring(PREFIX_LENGTH + 1, key.path.length - SUFFIX_LENGTH)) to value
            }.map { (key, value) ->
                key to jankson.fromJson(value, Book::class.java)
            }.toMap()
    }

    operator fun contains(id: Identifier): Boolean = books.contains(id)

    operator fun get(id: Identifier): Book =
        books.getOrElse(id) { throw IllegalArgumentException("Tried to get unknown book '$id' from BookManager") }
}
